File: Assets/Bundles/Server/Resources/TS/Services/Global/Entity/EntityService.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local _flamework_core = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init")
local Reflect = _flamework_core.Reflect
local Flamework = _flamework_core.Flamework
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ObjectUtil = require("Shared/rbxts_include/node_modules/@easy-games/unity-object-utils/init")
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local EntitySpawnEvent = require("Server/TS/Signals/EntitySpawnServerEvent").EntitySpawnEvent
local MoveCommandDataEvent = require("Server/TS/Signals/MoveCommandDataEvent").MoveCommandDataEvent
local CharacterEntity = require("Shared/TS/Entity/Character/CharacterEntity").CharacterEntity
local GameObjectBridge = require("Shared/TS/GameObjectBridge").GameObjectBridge
local Network = require("Shared/TS/Network").Network
local NetworkBridge = require("Shared/TS/NetworkBridge").NetworkBridge
local Task = require("Shared/TS/Util/Task").Task
local EntityPrefabType = require("Shared/TS/Entity/EntityPrefabType").EntityPrefabType
local EntityCommand = require("Server/TS/Services/Global/Entity/EntityCommand").EntityCommand
local EntityService
do
	EntityService = setmetatable({}, {
		__tostring = function()
			return "EntityService"
		end,
	})
	EntityService.__index = EntityService
	function EntityService.new(...)
		local self = setmetatable({}, EntityService)
		return self:constructor(...) or self
	end
	function EntityService:constructor(invService, chatService)
		self.invService = invService
		self.chatService = chatService
		self.idCounter = 1
		self.entities = {}
		self.loadedEntityPrefabs = {}
		self.chatService:RegisterCommand(EntityCommand.new())
		ServerSignals.MapLoad:connect(function(event)
			local pos = event.LoadedMap:GetMapSpawnPlatform()
			if pos then
				Task:Delay(1, function()
					local _fn = self
					local _exp = EntityPrefabType.HUMAN
					local _position = pos.Position
					local _vector3 = Vector3.new(-3, 2, 3)
					_fn:SpawnEntityForPlayer(nil, _exp, _position + _vector3)
				end)
			end
		end)
	end
	function EntityService:OnStart()
		(Flamework.resolveDependency("Bundles/Server/Services/Global/Player/PlayerService@PlayerService")):ObservePlayers(function(player)
			for _, entity in ObjectUtil.values(self.entities) do
				if TS.instanceof(entity, CharacterEntity) then
					local invDto = entity:GetInventory():Encode()
					Network.ServerToClient.UpdateInventory.Server:FireClient(player.clientId, invDto)
				end
			end
			local _exp = ObjectUtil.values(self.entities)
			local _arg0 = function(e)
				return e:Encode()
			end
			-- ▼ ReadonlyArray.map ▼
			local _newValue = table.create(#_exp)
			for _k, _v in _exp do
				_newValue[_k] = _arg0(_v, _k - 1, _exp)
			end
			-- ▲ ReadonlyArray.map ▲
			local dto = _newValue
			Network.ServerToClient.SpawnEntities.Server:FireClient(player.clientId, dto)
			return function()
				if player.Character then
					self:DespawnEntity(player.Character)
				end
			end
		end)
	end
	function EntityService:SpawnEntityForPlayer(player, entityPrefabType, pos, onDestroyed)
		local id = self.idCounter
		self.idCounter += 1
		local beforeEvent = ServerSignals.BeforeEntitySpawn:fire(id, player, pos or Vector3.new(0, 0, 0))
		-- Spawn character game object
		local entityPrefab
		local _loadedEntityPrefabs = self.loadedEntityPrefabs
		local _entityPrefabType = entityPrefabType
		if _loadedEntityPrefabs[_entityPrefabType] ~= nil then
			local _loadedEntityPrefabs_1 = self.loadedEntityPrefabs
			local _entityPrefabType_1 = entityPrefabType
			entityPrefab = _loadedEntityPrefabs_1[_entityPrefabType_1]
		else
			entityPrefab = AssetBridge:LoadAsset(entityPrefabType)
			if not entityPrefab then
				error("failed to find entity prefab: " .. entityPrefabType)
			end
			local _loadedEntityPrefabs_1 = self.loadedEntityPrefabs
			local _entityPrefabType_1 = entityPrefabType
			local _entityPrefab = entityPrefab
			_loadedEntityPrefabs_1[_entityPrefabType_1] = _entityPrefab
		end
		local entityGO = GameObjectBridge:InstantiateAt(entityPrefab, beforeEvent.spawnPosition, Quaternion.identity)
		entityGO.name = "entity_" .. tostring(id)
		local entityModelGO = entityGO.transform:Find("EntityModel")
		local destroyWatcher = entityGO:AddComponent("DestroyWatcher")
		destroyWatcher:OnDestroyedEvent(function()
			if onDestroyed ~= nil then
				onDestroyed()
			end
		end)
		if player then
			NetworkBridge:SpawnWithClientOwnership(entityGO, player.clientId)
		else
			NetworkBridge:Spawn(entityGO)
		end
		local nob = entityGO:GetComponent("NetworkObject")
		local inv = self.invService:MakeInventory()
		if player then
			self.invService:Subscribe(player.clientId, inv, true)
		end
		local _result = player
		if _result ~= nil then
			_result = _result.clientId
		end
		local entity = CharacterEntity.new(id, nob, _result, inv)
		self.entities[id] = entity
		if player then
			player.Character = entity
		end
		-- Spawn character model
		-- entity.SpawnCharacterModel(characterDef);
		local entityDriver = entityGO:GetComponent("EntityDriver")
		-- Custom move command data handling:
		entityDriver:OnDispatchCustomData(function(tick, customData)
			local allData = customData:Decode()
			for _, data in allData do
				local _result_1 = player
				if _result_1 ~= nil then
					_result_1 = _result_1.clientId
				end
				local _condition = _result_1
				if _condition == nil then
					_condition = -1
				end
				local moveEvent = MoveCommandDataEvent.new(_condition, tick, data.key, data.value)
				ServerSignals.CustomMoveCommand:Fire(moveEvent)
			end
		end)
		ServerSignals.EntitySpawn:Fire(EntitySpawnEvent.new(entity))
		Network.ServerToClient.SpawnEntities.Server:FireAllClients({ entity:Encode() })
		Network.ServerToClient.UpdateInventory.Server:FireAllClients(entity:GetInventory():Encode())
		entity:GetInventory():StartNetworkingDiffs()
		return entity
	end
	function EntityService:DespawnEntity(entity)
		ServerSignals.EntityDespawn:Fire(entity)
		entity:Destroy()
		local _entities = self.entities
		local _id = entity.id
		_entities[_id] = nil
	end
	function EntityService:GetEntityById(entityId)
		local _entities = self.entities
		local _entityId = entityId
		return _entities[_entityId]
	end
	function EntityService:GetEntityByClientId(clientId)
		local _exp = ObjectUtil.values(self.entities)
		local _arg0 = function(e)
			return e.ClientId == clientId
		end
		-- ▼ ReadonlyArray.find ▼
		local _result
		for _i, _v in _exp do
			if _arg0(_v, _i - 1, _exp) == true then
				_result = _v
				break
			end
		end
		-- ▲ ReadonlyArray.find ▲
		return _result
	end
	function EntityService:GetEntities()
		return ObjectUtil.values(self.entities)
	end
end
-- (Flamework) EntityService metadata
Reflect.defineMetadata(EntityService, "identifier", "Bundles/Server/Services/Global/Entity/EntityService@EntityService")
Reflect.defineMetadata(EntityService, "flamework:parameters", { "Bundles/Server/Services/Global/Inventory/InventoryService@InventoryService", "Bundles/Server/Services/Global/Chat/ChatService@ChatService" })
Reflect.defineMetadata(EntityService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(EntityService, "$:flamework@Service", Service, { {} })
return {
	EntityService = EntityService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Global/Entity/EntityService.ts
import { Dependency, OnStart, Service } from "@easy-games/flamework-core";
import ObjectUtil from "@easy-games/unity-object-utils";
import { ServerSignals } from "Server/ServerSignals";
import { EntitySpawnEvent } from "Server/Signals/EntitySpawnServerEvent";
import { MoveCommandDataEvent } from "Server/Signals/MoveCommandDataEvent";
import { CharacterEntity } from "Shared/Entity/Character/CharacterEntity";
import { Entity } from "Shared/Entity/Entity";
import { GameObjectUtil } from "Shared/GameObjectBridge";
import { Network } from "Shared/Network";
import { Player } from "Shared/Player/Player";
import { NetworkUtil } from "Shared/Util/NetworkUtil";
import { SignalPriority } from "Shared/Util/Signal";
import { EntityPrefabType } from "../../../Entity/EntityPrefabType";
import { ChatService } from "../Chat/ChatService";
import { InventoryService } from "../Inventory/InventoryService";
import { PlayerService } from "../Player/PlayerService";
import { EntityCommand } from "./EntityCommand";

@Service({})
export class EntityService implements OnStart {
	private idCounter = 1;
	private entities = new Map<number, Entity>();
	private loadedEntityPrefabs = new Map<EntityPrefabType, Object>();

	constructor(private readonly invService: InventoryService, private readonly chatService: ChatService) {
		this.chatService.RegisterCommand(new EntityCommand());
	}

	OnStart(): void {
		const playerService = Dependency<PlayerService>();
		playerService.ObservePlayers((player) => {
			for (let entity of ObjectUtil.values(this.entities)) {
				if (entity instanceof CharacterEntity) {
					const invDto = entity.GetInventory().Encode();
					Network.ServerToClient.UpdateInventory.Server.FireClient(player.clientId, invDto);
				}
			}
			const dto = ObjectUtil.values(this.entities).map((e) => e.Encode());
			Network.ServerToClient.SpawnEntities.Server.FireClient(player.clientId, dto);

			return () => {
				if (player.Character) {
					this.DespawnEntity(player.Character);
				}
			};
		}, SignalPriority.HIGHEST);
		ServerSignals.PlayerLeave.connect((event) => {
			if (event.player.Character) {
				this.DespawnEntity(event.player.Character);
			}
		});
	}

	public SpawnEntityForPlayer(
		player: Player | undefined,
		entityPrefabType: EntityPrefabType,
		pos?: Vector3,
		onDestroyed?: () => void,
	): Entity {
		const id = this.idCounter;
		this.idCounter++;

		const beforeEvent = ServerSignals.BeforeEntitySpawn.fire(id, player, pos ?? new Vector3(0, 0, 0));

		// Spawn character game object
		let entityPrefab: Object;
		if (this.loadedEntityPrefabs.has(entityPrefabType)) {
			entityPrefab = this.loadedEntityPrefabs.get(entityPrefabType)!;
		} else {
			entityPrefab = AssetBridge.LoadAsset(entityPrefabType);
			if (!entityPrefab) {
				throw "failed to find entity prefab: " + entityPrefabType;
			}
			this.loadedEntityPrefabs.set(entityPrefabType, entityPrefab);
		}
		const entityGO = GameObjectUtil.InstantiateAt(entityPrefab, beforeEvent.spawnPosition, Quaternion.identity);
		entityGO.name = `entity_${id}`;

		const entityModelGO = entityGO.transform.Find("EntityModel");

		const destroyWatcher = entityGO.AddComponent("DestroyWatcher") as DestroyWatcher;
		destroyWatcher.OnDestroyedEvent(() => {
			if (onDestroyed !== undefined) {
				onDestroyed();
			}
		});
		if (player) {
			NetworkUtil.SpawnWithClientOwnership(entityGO, player.clientId);
		} else {
			NetworkUtil.Spawn(entityGO);
		}

		const nob = entityGO.GetComponent("NetworkObject") as NetworkObject;

		const inv = this.invService.MakeInventory();
		if (player) {
			this.invService.Subscribe(player.clientId, inv, true);
		}
		const entity = new CharacterEntity(id, nob, player?.clientId, inv);
		this.entities.set(id, entity);
		if (player) {
			player.SetCharacter(entity);
		}

		// Spawn character model
		// entity.SpawnCharacterModel(characterDef);

		const entityDriver = entityGO.GetComponent<EntityDriver>();

		// Custom move command data handling:
		entityDriver.OnDispatchCustomData((tick, customData) => {
			const allData = customData.Decode() as { key: unknown; value: unknown }[];
			for (const data of allData) {
				const moveEvent = new MoveCommandDataEvent(player?.clientId ?? -1, tick, data.key, data.value);
				ServerSignals.CustomMoveCommand.Fire(moveEvent);
			}
		});

		// fire SpawnEntities after so the initial entity packet has all the latest info.
		ServerSignals.EntitySpawn.Fire(new EntitySpawnEvent(entity));

		Network.ServerToClient.SpawnEntities.Server.FireAllClients([entity.Encode()]);
		Network.ServerToClient.UpdateInventory.Server.FireAllClients(entity.GetInventory().Encode());
		entity.GetInventory().StartNetworkingDiffs();

		return entity;
	}

	public DespawnEntity(entity: Entity): void {
		ServerSignals.EntityDespawn.Fire(entity);
		entity.Destroy();
		this.entities.delete(entity.id);
	}

	public GetEntityById(entityId: number): Entity | undefined {
		return this.entities.get(entityId);
	}

	public GetEntityByClientId(clientId: number) {
		return ObjectUtil.values(this.entities).find((e) => e.ClientId === clientId);
	}

	public GetEntities(): Entity[] {
		return ObjectUtil.values(this.entities);
	}
}
